File: cmd/verification/verification.go
package verification

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/mailersend/mailersend-cli/internal/cmdutil"
	"github.com/mailersend/mailersend-cli/internal/output"
	"github.com/mailersend/mailersend-cli/internal/prompt"
	"github.com/mailersend/mailersend-cli/internal/sdkclient"
	"github.com/mailersend/mailersend-go"
	"github.com/spf13/cobra"
)

var Cmd = &cobra.Command{
	Use:   "verification",
	Short: "Email verification commands",
	Long:  "Verify individual email addresses and manage email verification lists.",
}

// --- Subcommand group for list operations ---

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "Manage verification lists",
	Long:  "List, create, verify, and inspect email verification lists.",
}

// --- init ---

func init() {
	// Single-email commands
	Cmd.AddCommand(verifyCmd)
	Cmd.AddCommand(verifyAsyncCmd)
	Cmd.AddCommand(statusCmd)

	// List subcommand group
	Cmd.AddCommand(listCmd)
	listCmd.AddCommand(listListCmd)
	listCmd.AddCommand(listGetCmd)
	listCmd.AddCommand(listCreateCmd)
	listCmd.AddCommand(listVerifyCmd)
	listCmd.AddCommand(listResultsCmd)

	// list list flags
	listListCmd.Flags().Int("limit", 0, "maximum number of lists to return (0 = all)")

	// list create flags
	listCreateCmd.Flags().String("name", "", "name for the verification list (required)")
	listCreateCmd.Flags().StringSlice("emails", nil, "comma-separated list of email addresses")
	listCreateCmd.Flags().String("emails-file", "", "path to file with one email per line")

	// list verify flags
	listVerifyCmd.Flags().Bool("wait", false, "poll until verification completes")

	// list results flags
	listResultsCmd.Flags().Int("limit", 0, "maximum number of results to return (0 = all)")
	listResultsCmd.Flags().String("status", "", "filter by status (valid, invalid, catch_all, mailbox_full, role, unknown)")
}

// --- Single-email commands ---

// verify -- uses raw HTTP because the SDK's VerifySingle only returns {status}
// but the API returns a richer response with email details.
var verifyCmd = &cobra.Command{
	Use:   "verify <email>",
	Short: "Verify a single email address",
	Args:  cobra.ExactArgs(1),
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		ctx := context.Background()
		payload, _ := json.Marshal(map[string]string{"email": args[0]})
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.mailersend.com/v1/email-verification/verify", bytes.NewReader(payload))
		if err != nil {
			return err
		}
		req.Header.Set("Authorization", "Bearer "+ms.APIKey())
		req.Header.Set("Accept", "application/json")
		req.Header.Set("Content-Type", "application/json")

		resp, err := ms.Client().Do(req)
		if err != nil {
			return sdkclient.WrapError(err)
		}
		defer resp.Body.Close() //nolint:errcheck

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("failed to read response: %w", err)
		}

		if resp.StatusCode >= 400 {
			return parseHTTPError(resp.StatusCode, body)
		}

		if cmdutil.JSONFlag(c) {
			var raw json.RawMessage
			if err := json.Unmarshal(body, &raw); err != nil {
				return err
			}
			return output.JSON(raw)
		}

		var respData struct {
			Data struct {
				Email  json.RawMessage `json:"email"`
				Status string          `json:"status"`
			} `json:"data"`
		}
		if err := json.Unmarshal(body, &respData); err != nil {
			return fmt.Errorf("failed to parse response: %w", err)
		}

		headers := []string{"FIELD", "VALUE"}
		rows := [][]string{
			{"Email", args[0]},
			{"Status", respData.Data.Status},
		}

		// Try to extract additional email info
		if respData.Data.Email != nil {
			var emailInfo map[string]interface{}
			if err := json.Unmarshal(respData.Data.Email, &emailInfo); err == nil {
				for _, key := range []string{"local_part", "domain", "mx_found", "mx_record"} {
					if v, ok := emailInfo[key]; ok && v != nil {
						rows = append(rows, []string{key, fmt.Sprintf("%v", v)})
					}
				}
			}
		}

		output.Table(headers, rows)
		return nil
	},
}

// verify-async -- uses raw HTTP since the SDK doesn't have this endpoint.
var verifyAsyncCmd = &cobra.Command{
	Use:   "verify-async <email>",
	Short: "Verify a single email address asynchronously",
	Args:  cobra.ExactArgs(1),
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		ctx := context.Background()
		payload, _ := json.Marshal(map[string]string{"email": args[0]})
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.mailersend.com/v1/email-verification/verify-async", bytes.NewReader(payload))
		if err != nil {
			return err
		}
		req.Header.Set("Authorization", "Bearer "+ms.APIKey())
		req.Header.Set("Accept", "application/json")
		req.Header.Set("Content-Type", "application/json")

		resp, err := ms.Client().Do(req)
		if err != nil {
			return sdkclient.WrapError(err)
		}
		defer resp.Body.Close() //nolint:errcheck

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("failed to read response: %w", err)
		}

		if resp.StatusCode >= 400 {
			return parseHTTPError(resp.StatusCode, body)
		}

		if cmdutil.JSONFlag(c) {
			var raw json.RawMessage
			if err := json.Unmarshal(body, &raw); err != nil {
				return err
			}
			return output.JSON(raw)
		}

		var respData struct {
			Data struct {
				ID      string `json:"id"`
				Address string `json:"address"`
				Status  string `json:"status"`
			} `json:"data"`
		}
		if err := json.Unmarshal(body, &respData); err != nil {
			return fmt.Errorf("failed to parse response: %w", err)
		}

		headers := []string{"FIELD", "VALUE"}
		rows := [][]string{
			{"ID", respData.Data.ID},
			{"Address", respData.Data.Address},
			{"Status", respData.Data.Status},
		}

		output.Table(headers, rows)
		return nil
	},
}

// status -- uses raw HTTP since the SDK doesn't have this endpoint.
var statusCmd = &cobra.Command{
	Use:   "status <id>",
	Short: "Get async email verification status",
	Args:  cobra.ExactArgs(1),
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		ctx := context.Background()
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.mailersend.com/v1/email-verification/verify-async/"+args[0], nil)
		if err != nil {
			return err
		}
		req.Header.Set("Authorization", "Bearer "+ms.APIKey())
		req.Header.Set("Accept", "application/json")

		resp, err := ms.Client().Do(req)
		if err != nil {
			return sdkclient.WrapError(err)
		}
		defer resp.Body.Close() //nolint:errcheck

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("failed to read response: %w", err)
		}

		if resp.StatusCode >= 400 {
			return parseHTTPError(resp.StatusCode, body)
		}

		if cmdutil.JSONFlag(c) {
			var raw json.RawMessage
			if err := json.Unmarshal(body, &raw); err != nil {
				return err
			}
			return output.JSON(raw)
		}

		var respData struct {
			Data struct {
				ID      string          `json:"id"`
				Address string          `json:"address"`
				Status  string          `json:"status"`
				Result  json.RawMessage `json:"result"`
				Error   json.RawMessage `json:"error"`
			} `json:"data"`
		}
		if err := json.Unmarshal(body, &respData); err != nil {
			return fmt.Errorf("failed to parse response: %w", err)
		}

		headers := []string{"FIELD", "VALUE"}
		rows := [][]string{
			{"ID", respData.Data.ID},
			{"Address", respData.Data.Address},
			{"Status", respData.Data.Status},
		}

		if respData.Data.Result != nil && string(respData.Data.Result) != "null" {
			rows = append(rows, []string{"Result", string(respData.Data.Result)})
		}
		if respData.Data.Error != nil && string(respData.Data.Error) != "null" {
			rows = append(rows, []string{"Error", string(respData.Data.Error)})
		}

		output.Table(headers, rows)
		return nil
	},
}

// --- List subcommands ---

// list list
var listListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all verification lists",
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		limit, _ := c.Flags().GetInt("limit")

		ctx := context.Background()
		items, err := sdkclient.FetchAll(ctx, func(ctx context.Context, page, perPage int) ([]mailersend.EmailVerification, bool, error) {
			root, _, err := ms.EmailVerification.List(ctx, &mailersend.ListEmailVerificationOptions{
				Page:  page,
				Limit: perPage,
			})
			if err != nil {
				return nil, false, sdkclient.WrapError(err)
			}
			return root.Data, root.Links.Next != "", nil
		}, limit)
		if err != nil {
			return err
		}

		if cmdutil.JSONFlag(c) {
			return output.JSON(items)
		}

		headers := []string{"ID", "NAME", "TOTAL", "STATUS", "CREATED AT"}
		var rows [][]string
		for _, item := range items {
			statusName := ""
			if item.Status.Name != "" {
				statusName = item.Status.Name
			}
			createdAt := ""
			if !item.CreatedAt.IsZero() {
				createdAt = item.CreatedAt.Format("2006-01-02 15:04:05")
			}

			rows = append(rows, []string{
				item.Id,
				item.Name,
				strconv.Itoa(item.Total),
				statusName,
				createdAt,
			})
		}

		output.Table(headers, rows)
		return nil
	},
}

// list get
var listGetCmd = &cobra.Command{
	Use:   "get <id>",
	Short: "Get verification list details",
	Args:  cobra.ExactArgs(1),
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		ctx := context.Background()
		result, _, err := ms.EmailVerification.Get(ctx, args[0])
		if err != nil {
			return sdkclient.WrapError(err)
		}

		if cmdutil.JSONFlag(c) {
			return output.JSON(result)
		}

		d := result.Data
		statusName := ""
		if d.Status.Name != "" {
			statusName = d.Status.Name
		}

		createdAt := ""
		if !d.CreatedAt.IsZero() {
			createdAt = d.CreatedAt.Format("2006-01-02 15:04:05")
		}
		updatedAt := ""
		if !d.UpdatedAt.IsZero() {
			updatedAt = d.UpdatedAt.Format("2006-01-02 15:04:05")
		}

		verificationStarted := fmt.Sprintf("%v", d.VerificationStarted)
		verificationEnded := fmt.Sprintf("%v", d.VerificationEnded)
		if verificationStarted == "<nil>" {
			verificationStarted = ""
		}
		if verificationEnded == "<nil>" {
			verificationEnded = ""
		}

		stats := d.Statistics

		headers := []string{"FIELD", "VALUE"}
		rows := [][]string{
			{"ID", d.Id},
			{"Name", d.Name},
			{"Total", strconv.Itoa(d.Total)},
			{"Status", statusName},
			{"Source", d.Source},
			{"Verification Started", verificationStarted},
			{"Verification Ended", verificationEnded},
			{"Created At", createdAt},
			{"Updated At", updatedAt},
			{"", ""},
			{"--- Statistics ---", ""},
			{"Valid", strconv.Itoa(stats.Valid)},
			{"Catch All", strconv.Itoa(stats.CatchAll)},
			{"Mailbox Full", strconv.Itoa(stats.MailboxFull)},
			{"Role Based", strconv.Itoa(stats.RoleBased)},
			{"Unknown", strconv.Itoa(stats.Unknown)},
			{"Syntax Error", strconv.Itoa(stats.SyntaxError)},
			{"Typo", strconv.Itoa(stats.Typo)},
			{"Mailbox Not Found", strconv.Itoa(stats.MailboxNotFound)},
			{"Disposable", strconv.Itoa(stats.Disposable)},
			{"Mailbox Blocked", strconv.Itoa(stats.MailboxBlocked)},
			{"Failed", strconv.Itoa(stats.Failed)},
		}

		output.Table(headers, rows)
		return nil
	},
}

// list create
var listCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a verification list",
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		name, _ := c.Flags().GetString("name")
		name, err = prompt.RequireArg(name, "name", "Verification list name")
		if err != nil {
			return err
		}
		emails, _ := c.Flags().GetStringSlice("emails")
		emailsFile, _ := c.Flags().GetString("emails-file")

		if emailsFile != "" {
			data, err := os.ReadFile(emailsFile)
			if err != nil {
				return fmt.Errorf("failed to read emails file: %w", err)
			}
			lines := strings.Split(string(data), "\n")
			for _, line := range lines {
				trimmed := strings.TrimSpace(line)
				if trimmed != "" {
					emails = append(emails, trimmed)
				}
			}
		}

		if len(emails) == 0 {
			return fmt.Errorf("provide emails via --emails or --emails-file")
		}

		ctx := context.Background()
		result, _, err := ms.EmailVerification.Create(ctx, &mailersend.CreateEmailVerificationOptions{
			Name:   name,
			Emails: emails,
		})
		if err != nil {
			return sdkclient.WrapError(err)
		}

		if cmdutil.JSONFlag(c) {
			return output.JSON(result)
		}

		output.Success(fmt.Sprintf("Verification list created: %s (ID: %s)", result.Data.Name, result.Data.Id))
		return nil
	},
}

// list verify
var listVerifyCmd = &cobra.Command{
	Use:   "verify <id>",
	Short: "Start verification of a list",
	Args:  cobra.ExactArgs(1),
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		id := args[0]
		ctx := context.Background()

		result, _, err := ms.EmailVerification.Verify(ctx, id)
		if err != nil {
			return sdkclient.WrapError(err)
		}

		wait, _ := c.Flags().GetBool("wait")

		if !wait {
			if cmdutil.JSONFlag(c) {
				return output.JSON(result)
			}

			output.Success(fmt.Sprintf("Verification started for list %s.", id))
			return nil
		}

		// Poll until done
		for {
			time.Sleep(5 * time.Second)

			pollResult, _, err := ms.EmailVerification.Get(ctx, id)
			if err != nil {
				return sdkclient.WrapError(err)
			}

			statusName := ""
			if pollResult.Data.Status.Name != "" {
				statusName = pollResult.Data.Status.Name
			}

			fmt.Printf("Waiting... (status: %s)\n", statusName)

			if statusName == "verified" || statusName == "failed" {
				if cmdutil.JSONFlag(c) {
					return output.JSON(pollResult)
				}

				if statusName == "verified" {
					output.Success(fmt.Sprintf("Verification completed for list %s.", id))
				} else {
					output.Error(fmt.Sprintf("Verification failed for list %s.", id))
				}
				return nil
			}
		}
	},
}

// list results
var listResultsCmd = &cobra.Command{
	Use:   "results <id>",
	Short: "Get verification results for a list",
	Args:  cobra.ExactArgs(1),
	RunE: func(c *cobra.Command, args []string) error {
		ms, err := cmdutil.NewSDKClient(c)
		if err != nil {
			return err
		}

		id := args[0]
		limit, _ := c.Flags().GetInt("limit")

		ctx := context.Background()
		items, err := sdkclient.FetchAll(ctx, func(ctx context.Context, page, perPage int) ([]mailersend.Result, bool, error) {
			root, _, err := ms.EmailVerification.GetResults(ctx, &mailersend.GetEmailVerificationOptions{
				EmailVerificationId: id,
				Page:                page,
				Limit:               perPage,
			})
			if err != nil {
				return nil, false, sdkclient.WrapError(err)
			}
			return root.Data, root.Links.Next != "", nil
		}, limit)
		if err != nil {
			return err
		}

		if cmdutil.JSONFlag(c) {
			return output.JSON(items)
		}

		headers := []string{"EMAIL", "RESULT", "REASON"}
		var rows [][]string
		for _, item := range items {
			email := item.Address
			resultStr := item.Result
			reason := ""
			rows = append(rows, []string{email, resultStr, reason})
		}

		output.Table(headers, rows)
		return nil
	},
}

// --- Helpers ---

// parseHTTPError creates a CLIError from a raw HTTP error response.
func parseHTTPError(statusCode int, body []byte) error {
	cliErr := &sdkclient.CLIError{
		StatusCode: statusCode,
	}
	if len(body) > 0 {
		var parsed struct {
			Message string              `json:"message"`
			Errors  map[string][]string `json:"errors"`
		}
		if json.Unmarshal(body, &parsed) == nil {
			cliErr.Message = parsed.Message
			if len(parsed.Errors) > 0 {
				cliErr.Errors = parsed.Errors
			}
		}
		if cliErr.Message == "" {
			cliErr.Message = string(body)
		}
	}
	return cliErr
}

File: src/commands/verification.rs
use anyhow::{bail, Context, Result};
use clap::{Args, Subcommand};
use serde_json::{json, Value};

use crate::api;
use crate::cli::Ctx;
use crate::output;
use crate::prompt;
use crate::util::{jint, jpath, jstr};

#[derive(Args)]
pub struct Cmd {
    #[command(subcommand)]
    command: Sub,
}

#[derive(Subcommand)]
enum Sub {
    /// Verify a single email address
    #[command(arg_required_else_help = true)]
    Verify {
        /// email address to verify
        email: String,
    },
    /// Verify a single email address asynchronously
    #[command(arg_required_else_help = true)]
    VerifyAsync {
        /// email address to verify
        email: String,
    },
    /// Get async email verification status
    #[command(arg_required_else_help = true)]
    Status {
        /// async verification ID
        id: String,
    },
    /// Manage verification lists
    #[command(long_about = "List, create, verify, and inspect email verification lists.")]
    List {
        #[command(subcommand)]
        command: ListSub,
    },
}

#[derive(Subcommand)]
enum ListSub {
    /// List all verification lists
    List {
        /// maximum number of lists to return (0 = all)
        #[arg(long, default_value_t = 0)]
        limit: u64,
    },
    /// Get verification list details
    #[command(arg_required_else_help = true)]
    Get {
        /// verification list ID
        id: String,
    },
    /// Create a verification list
    Create {
        /// name for the verification list (required)
        #[arg(long, default_value = "")]
        name: String,
        /// comma-separated list of email addresses
        #[arg(long, value_delimiter = ',')]
        emails: Option<Vec<String>>,
        /// path to file with one email per line
        #[arg(long, default_value = "")]
        emails_file: String,
    },
    /// Start verification of a list
    #[command(arg_required_else_help = true)]
    Verify {
        /// verification list ID
        id: String,
        /// poll until verification completes
        #[arg(long)]
        wait: bool,
    },
    /// Get verification results for a list
    #[command(arg_required_else_help = true)]
    Results {
        /// verification list ID
        id: String,
        /// maximum number of results to return (0 = all)
        #[arg(long, default_value_t = 0)]
        limit: u64,
        /// filter by status (valid, invalid, catch_all, mailbox_full, role, unknown)
        #[arg(long, default_value = "")]
        status: String,
    },
}

pub fn run(ctx: &Ctx, cmd: Cmd) -> Result<()> {
    match cmd.command {
        Sub::Verify { email } => verify(ctx, &email),
        Sub::VerifyAsync { email } => verify_async(ctx, &email),
        Sub::Status { id } => status(ctx, &id),
        Sub::List { command } => match command {
            ListSub::List { limit } => list_list(ctx, limit),
            ListSub::Get { id } => list_get(ctx, &id),
            ListSub::Create {
                name,
                emails,
                emails_file,
            } => list_create(ctx, &name, emails, &emails_file),
            ListSub::Verify { id, wait } => list_verify(ctx, &id, wait),
            ListSub::Results {
                id,
                limit,
                status: _,
            } => list_results(ctx, &id, limit),
        },
    }
}

fn fmt_datetime(s: &str) -> String {
    chrono::DateTime::parse_from_rfc3339(s)
        .map(|d| d.format("%Y-%m-%d %H:%M:%S").to_string())
        .unwrap_or_default()
}

fn fmt_value(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

fn verify(ctx: &Ctx, email: &str) -> Result<()> {
    let client = ctx.client()?;

    let body = client.post("/email-verification/verify", &json!({ "email": email }))?;

    if ctx.json {
        return output::json(&body);
    }

    let data = body.get("data").cloned().unwrap_or(Value::Null);

    let mut rows = vec![
        vec!["Email".to_string(), email.to_string()],
        vec!["Status".to_string(), jstr(&data, "status")],
    ];

    if let Some(info) = data.get("email").and_then(Value::as_object) {
        for key in ["local_part", "domain", "mx_found", "mx_record"] {
            if let Some(v) = info.get(key) {
                if !v.is_null() {
                    rows.push(vec![key.to_string(), fmt_value(v)]);
                }
            }
        }
    }

    output::table(&["FIELD", "VALUE"], &rows);
    Ok(())
}

fn verify_async(ctx: &Ctx, email: &str) -> Result<()> {
    let client = ctx.client()?;

    let body = client.post(
        "/email-verification/verify-async",
        &json!({ "email": email }),
    )?;

    if ctx.json {
        return output::json(&body);
    }

    let data = body.get("data").cloned().unwrap_or(Value::Null);

    let rows = vec![
        vec!["ID".to_string(), jstr(&data, "id")],
        vec!["Address".to_string(), jstr(&data, "address")],
        vec!["Status".to_string(), jstr(&data, "status")],
    ];

    output::table(&["FIELD", "VALUE"], &rows);
    Ok(())
}

fn status(ctx: &Ctx, id: &str) -> Result<()> {
    let client = ctx.client()?;

    let body = client.get(&format!("/email-verification/verify-async/{id}"))?;

    if ctx.json {
        return output::json(&body);
    }

    let data = body.get("data").cloned().unwrap_or(Value::Null);

    let mut rows = vec![
        vec!["ID".to_string(), jstr(&data, "id")],
        vec!["Address".to_string(), jstr(&data, "address")],
        vec!["Status".to_string(), jstr(&data, "status")],
    ];

    if let Some(result) = data.get("result") {
        if !result.is_null() {
            rows.push(vec!["Result".to_string(), result.to_string()]);
        }
    }
    if let Some(error) = data.get("error") {
        if !error.is_null() {
            rows.push(vec!["Error".to_string(), error.to_string()]);
        }
    }

    output::table(&["FIELD", "VALUE"], &rows);
    Ok(())
}

fn list_list(ctx: &Ctx, limit: u64) -> Result<()> {
    let client = ctx.client()?;

    let items = api::fetch_all_paged(&client, "/email-verification", &[], limit)?;

    if ctx.json {
        return output::json(&items);
    }

    let rows: Vec<Vec<String>> = items
        .iter()
        .map(|item| {
            vec![
                jstr(item, "id"),
                jstr(item, "name"),
                jint(item, "total").to_string(),
                jpath(item, "status.name"),
                fmt_datetime(&jstr(item, "created_at")),
            ]
        })
        .collect();

    output::table(&["ID", "NAME", "TOTAL", "STATUS", "CREATED AT"], &rows);
    Ok(())
}

fn list_get(ctx: &Ctx, id: &str) -> Result<()> {
    let client = ctx.client()?;

    let body = client.get(&format!("/email-verification/{id}"))?;

    if ctx.json {
        return output::json(&body);
    }

    let d = body.get("data").cloned().unwrap_or(Value::Null);

    let opt_value = |key: &str| -> String {
        match d.get(key) {
            None | Some(Value::Null) => String::new(),
            Some(v) => fmt_value(v),
        }
    };

    let stats = d.get("statistics").cloned().unwrap_or(Value::Null);

    let rows = vec![
        vec!["ID".to_string(), jstr(&d, "id")],
        vec!["Name".to_string(), jstr(&d, "name")],
        vec!["Total".to_string(), jint(&d, "total").to_string()],
        vec!["Status".to_string(), jpath(&d, "status.name")],
        vec!["Source".to_string(), jstr(&d, "source")],
        vec![
            "Verification Started".to_string(),
            opt_value("verification_started"),
        ],
        vec![
            "Verification Ended".to_string(),
            opt_value("verification_ended"),
        ],
        vec![
            "Created At".to_string(),
            fmt_datetime(&jstr(&d, "created_at")),
        ],
        vec![
            "Updated At".to_string(),
            fmt_datetime(&jstr(&d, "updated_at")),
        ],
        vec![String::new(), String::new()],
        vec!["--- Statistics ---".to_string(), String::new()],
        vec!["Valid".to_string(), jint(&stats, "valid").to_string()],
        vec![
            "Catch All".to_string(),
            jint(&stats, "catch_all").to_string(),
        ],
        vec![
            "Mailbox Full".to_string(),
            jint(&stats, "mailbox_full").to_string(),
        ],
        vec![
            "Role Based".to_string(),
            jint(&stats, "role_based").to_string(),
        ],
        vec!["Unknown".to_string(), jint(&stats, "unknown").to_string()],
        vec![
            "Syntax Error".to_string(),
            jint(&stats, "syntax_error").to_string(),
        ],
        vec!["Typo".to_string(), jint(&stats, "typo").to_string()],
        vec![
            "Mailbox Not Found".to_string(),
            jint(&stats, "mailbox_not_found").to_string(),
        ],
        vec![
            "Disposable".to_string(),
            jint(&stats, "disposable").to_string(),
        ],
        vec![
            "Mailbox Blocked".to_string(),
            jint(&stats, "mailbox_blocked").to_string(),
        ],
        vec!["Failed".to_string(), jint(&stats, "failed").to_string()],
    ];

    output::table(&["FIELD", "VALUE"], &rows);
    Ok(())
}

fn list_create(
    ctx: &Ctx,
    name: &str,
    emails: Option<Vec<String>>,
    emails_file: &str,
) -> Result<()> {
    let client = ctx.client()?;

    let name = prompt::require_arg(name, "name", "Verification list name")?;

    let mut emails = emails.unwrap_or_default();
    if !emails_file.is_empty() {
        let data = std::fs::read_to_string(emails_file).context("failed to read emails file")?;
        for line in data.lines() {
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                emails.push(trimmed.to_string());
            }
        }
    }

    if emails.is_empty() {
        bail!("provide emails via --emails or --emails-file");
    }

    let result = client.post(
        "/email-verification",
        &json!({ "name": name, "emails": emails }),
    )?;

    if ctx.json {
        return output::json(&result);
    }

    let data = result.get("data").cloned().unwrap_or(Value::Null);
    output::success(&format!(
        "Verification list created: {} (ID: {})",
        jstr(&data, "name"),
        jstr(&data, "id")
    ));
    Ok(())
}

fn list_verify(ctx: &Ctx, id: &str, wait: bool) -> Result<()> {
    let client = ctx.client()?;

    let result = client.get(&format!("/email-verification/{id}/verify"))?;

    if !wait {
        if ctx.json {
            return output::json(&result);
        }

        output::success(&format!("Verification started for list {id}."));
        return Ok(());
    }

    loop {
        std::thread::sleep(std::time::Duration::from_secs(5));

        let poll = client.get(&format!("/email-verification/{id}"))?;
        let status_name = jpath(&poll, "data.status.name");

        println!("Waiting... (status: {status_name})");

        if status_name == "verified" || status_name == "failed" {
            if ctx.json {
                return output::json(&poll);
            }

            if status_name == "verified" {
                output::success(&format!("Verification completed for list {id}."));
            } else {
                output::error(&format!("Verification failed for list {id}."));
            }
            return Ok(());
        }
    }
}

fn list_results(ctx: &Ctx, id: &str, limit: u64) -> Result<()> {
    let client = ctx.client()?;

    let items = api::fetch_all_paged(
        &client,
        &format!("/email-verification/{id}/results"),
        &[],
        limit,
    )?;

    if ctx.json {
        return output::json(&items);
    }

    let rows: Vec<Vec<String>> = items
        .iter()
        .map(|item| vec![jstr(item, "address"), jstr(item, "result"), String::new()])
        .collect();

    output::table(&["EMAIL", "RESULT", "REASON"], &rows);
    Ok(())
}
